### PR TITLE
Make scalar multiplication type stable

### PR DESCRIPTION
**Pull request by eschnett**
Also provide a unary + operator.

Closes https://github.com/JuliaArrays/StaticArrays.jl/issues/899.

### CODE REVIEW

**Comment by mateuszbaran** [comment]
Thanks for a patch, could you also add appropriate `@inferred` tests for this case?

**Comment by eschnett** [comment]
I see errors
```
Test Failed at /Users/eschnett/.julia/dev/StaticArrays/test/ambiguities.jl:19
  Expression: length(detect_ambiguities(Base, LinearAlgebra, StaticArrays)) <= allowable_ambiguities
   Evaluated: 61 <= 0
ERROR: LoadError: LoadError: There was an error during testing
in expression starting at /Users/eschnett/.julia/dev/StaticArrays/test/ambiguities.jl:13
in expression starting at /Users/eschnett/.julia/dev/StaticArrays/test/runtests.jl:38
ERROR: Package StaticArrays errored during testing
```

However, I already see them with the head of the master branch, it's not caused by my pull request.

**Comment by mateuszbaran** [comment]
These errors are a separate issue.

File: test/linalg.jl [hunk 1]
```diff
@@ -36,6 +39,13 @@ StaticArrays.similar_type(::Union{RotMat2,Type{RotMat2}}) = SMatrix{2,2,Float64,
         #@test @inferred(v3 + v2) === @SVector [6, 7, 8, 9]
         #@test @inferred(v1 - v4) === @SVector [-2, 1, 4, 7]
         #@test @inferred(v3 - v2) === @SVector [-2, 1, 4, 7]
+
+        if VERSION ≥ v"1.2"
+            # #899 matrix-of-matrix
+            A = SMatrix{1,1}([1])
+            B = SMatrix{1,1}([A])
+            @test @inferred(1.0 * B) === SMatrix{1, 1, SMatrix{1, 1, Float64, 1}, 1}(B)
```
Comment (mateuszbaran) [inline test/linalg.jl:47]: Could you also add `@inferred` tests for the other three operations (`B*1.0`, `1.0 \ B`, `B / 1.0`)?

Comment (eschnett) [reply test/linalg.jl:47]: Done.

**Review by mateuszbaran — APPROVED** [review]

**Comment by mateuszbaran** [comment]
One last thing, could you bump version to 1.1.2?

**Pull request merged**